app/celery/process_delivery_status_result_tasks.py

```diff
@@ -1,13 +1,296 @@
+from app.config import QueueNames
+from app.dao.service_callback_dao import dao_get_callback_include_payload_status
+
+from app.models import (
+    NOTIFICATION_DELIVERED,
+    NOTIFICATION_TECHNICAL_FAILURE,
+    NOTIFICATION_PERMANENT_FAILURE,
+    Notification,
+    NOTIFICATION_PREFERENCES_DECLINED,
+)
+
+FINAL_STATUS_STATES = [
+    NOTIFICATION_DELIVERED,
+    NOTIFICATION_PERMANENT_FAILURE,
+    NOTIFICATION_TECHNICAL_FAILURE,
+    NOTIFICATION_PREFERENCES_DECLINED,
+]
+
 
 # Create SQS Queue for Process Deliver Status.
-@notify_celery.task(bind=True, name="process-delivery-status-result", max_retries=48, default_retry_delay=300)
+@notify_celery.task(bind=True, name="process-delivery-status-result", max_retries=48, default_retry_delay=300, )
+@statsd(namespace="tasks")
 def process_delivery_status(self, event: CeleryEvent) -> bool:
-    current_app.logger.info('processing delivery status: %s', event)
-    from app import clients
-    provider = clients.get_sms_client('twilio')
-    current_app.logger.info('Translated message: %s', provider.translate_delivery_status(event['message']))
+
+    """Celery task for updating the delivery status of a notification"""
+
+    # preset variables to address "unbounded local variable"
+    sqs_message = None
+    notification_platform_status = None
+
+    # log that we are processing the delivery status
+    current_app.logger.info("processing delivery status: %s", event)
+
+    # first attempt to process the incoming event
+    sqs_message = _get_sqs_message(self, event)
+
+    # get the provider
+    (provider_name, provider) = _get_provider_info(self, sqs_message)
+
+    body = sqs_message.get("body")
+    current_app.logger.info("retrieved delivery status body: %s", body)
+
+    # get notification_platform_status
+    notification_platform_status = _get_notification_platform_status(self, provider, body, sqs_message)
+
+    # get parameters from notification platform status
+    (payload, reference, notification_status,
+     number_of_message_parts, price_in_millicents_usd) = _get_notification_parameters(notification_platform_status)
+
+    # retrieves the inbound message for this provider we are updating the status of the outbound message
+    notification, should_retry, should_exit = attempt_to_get_notification(
+        reference, notification_status, str(time.time() * 1000)
+    )
+
+    # the race condition scenario if we got the delivery status before we actually record the sms
+    if should_retry or (notification is None):
+        self.retry(queue=QueueNames.RETRY)
+
+    if should_exit:
+        return False
+
+    try:
+        # calculate pricing
+        _calculate_pricing(price_in_millicents_usd, notification, notification_status, number_of_message_parts)
+        current_app.logger.info(
+            "Delivery Status callback return status of %s for notification: %s",
+            notification_status,
+            notification.id,
+        )
```

Comment:
This logging so out of place. It doesn't have anything to do with price calculations. We could move this out of the try or put it above the calculation if we wanted to be safe.